WalletWasabi.Documentation/EventSourcing.md

```diff
@@ -0,0 +1,119 @@
+  * For eventually consistent sequence of operations accross bounded contexts there is a concept of [Saga](#saga)
+
+### Command
+
+* Command is an input for a state transition of an [Aggregate](#aggregate).
+* Command is implemented as an immutable serializable value object.
+* Command is received and processed by the [Command-Processor](#command-processor).
+* Command needs to contain [IdempotenceId](#idempotenceid) provided by the orignator of the command (client) to implement at-most-once work around of The Two Generals' Problem.
+  * Second command with the same [IdempotenceId](#idempotenceid) is simply silently ignored.
+* Issuing a command to a [Command-Processor](#command-processor) is the only legal way of transitioning state of an [Aggregate](#aggregate) and thus generating new [Events](#event).
+* Commands are not persisted by default. Just events as a result of a successful command are persisted.
+* Command can fail. In such case an originator is informed by an error result of the command. However command failure is not broadcasted to the [PubSub Bus](#pubsub-bus) for general audience because state transition didn't happen so there is nothing to broadcast. And command failure is not persisted by default.
+
+### Command-Processor
+
+* Command-Processor transitions state of an aggregate based on the [Command](#command) as an input and current internal state of the [Aggregate](#aggregate) by producing new [Events](#event).
+* Mathematically it is a state machine transition rule.
+* Command-Processor implementation must not have any side-effects. Command-Processor needs to be deterministic. The only input is [Command](#command) and internal [Aggregate](#aggregate) state and only output are new [Events](#event). Any external side-efect needs to be implemented externally as an reaction to a generated event by subscribing to a [PubSub Bus](#pubsub-bus).
+  * One of the practical reasons for banning side-effects is because command can be retried again in case of an optimistic concurrency conflict detected upon persisting events.
+* Command-Processor implementation can assume it is a single threaded and strongly sequentially serializable. Practical implementation is usually optimistic concurrency strategy where command is retried again on an up to date version of an [Aggregate](#aggregate) after conflict has been detected during persisting of [Events](#event).
+
+### Read-Model
+
+* Read-model handles all read-only clients' requests.
+* Read-model is [Eventually consistent](#eventual-consistency). That means its state might be transitionally slightly out of date when compared to [Event Store](#event-store).
+* Read-model has an internal copy of all the data. All queries are directly resolved from internal read-model state without touching the [Event Store](#event-store).
```

Comment:
this is not strictly a requirement from the implementation PoV, right?

Reply:
Well, it is the standard clean implementation. Industry is doing Event Sourcing for performance and cost reasons. Event Store is expensive redundant database that needs back ups and what not. Whereas Read Model is cheap cache that can be recreated on demand so you can use 10x cheaper storage for Read Models. So it is usually much cheaper to have redundant Read Models than to put more strain to your.

It is actually independent concept called CQRS Command Query Responsibility Segregation https://martinfowler.com/bliki/CQRS.html I suppose you can do Event Sourcing without CQRS but in reality nobody is doing it and they come together as a default expectation AFAIK.

Reply:
IMO it is a good default without adding to much complication

Reply:
I suppose it is good to explicitly state that redundancy is expected and welcome because for some reason we still avoid redundancy for the 1 KiB of memory in the 60s reasons.

Reply:
Basically you can write another whole document about Read-Models. I wanted to simplify it as much as possible with directionally right ideas while omitting the details. Kind of the lie for the children 😅